clients/src/test/java/org/apache/kafka/test/TestUtils.java

```diff
@@ -225,21 +227,28 @@ public static Properties consumerConfig(final String bootstrapServers, Class key
     }
 
     /**
-     *  uses default value of 30 seconds for timeout
+     *  uses default value of 15 seconds for timeout, and default message for failure
      */
-    public static void waitForCondition(TestCondition testCondition) throws InterruptedException {
-        waitForCondition(testCondition, 30000);
+    public static void waitForCondition(TestCondition testCondition, String conditionDetails) throws InterruptedException {
+        waitForCondition(testCondition, 15000, conditionDetails);
     }
 
     /**
-     *  Used to wait for specific conditions/state to be me during a test
-     *  this is meant to be a replacement for using Thread.sleep
+     * Wait for condition to be met for at most {@code maxWaitMs} and throw assertion failure otherwise.
+     * This should be used instead of {@code Thread.sleep} whenever possible as it allows a longer timeout to be used
+     * without unnecessarily increasing test time (as the condition is checked frequently). The longer timeout is needed to
+     * avoid transient failures due to slow or overloaded machines.
      */
-    public static void waitForCondition(TestCondition testCondition, long maxTimeMillis) throws InterruptedException {
+    public static void waitForCondition(TestCondition testCondition, long maxWaitMs, String conditionDetails) throws InterruptedException {
         long startTime = System.currentTimeMillis();
 
-        while (!testCondition.conditionMet() && ((System.currentTimeMillis() - startTime) < maxTimeMillis)) {
-            Thread.sleep(Math.min(maxTimeMillis, 100L));
+
+        while (!testCondition.conditionMet() && ((System.currentTimeMillis() - startTime) < maxWaitMs)) {
+            Thread.sleep(Math.min(maxWaitMs, 100L));
+        }
+
+        if (!testCondition.conditionMet() && conditionDetails != null) {
```

Comment:
Given our documentation, I think we should remove the `conditionDetails != null` check. And instead we should convert null to an empty string so that the error message doesn't look weird. What do you think?